src/extra/gpkg/read.jl

```diff
@@ -0,0 +1,217 @@
+# ------------------------------------------------------------------
+# Licensed under the MIT License. See LICENSE in the project root.
+# ------------------------------------------------------------------
+
+function gpkgread(fname; layer=1)
+  db = gpkgdatabase(fname)
+  table, geoms = gpkgextract(db; layer)
+  DBInterface.close!(db)
+  georef(table, geoms)
+end
+
+function gpkgdatabase(fname)
+  # connect to SQLite database on disk
+  db = SQLite.DB(fname)
+
+  # According to https://www.geopackage.org/spec/#r6 and https://www.geopackage.org/spec/#r7
+  # PRAGMA integrity_check returns a single row with the value 'ok'
+  # PRAGMA foreign_key_check (w/ no parameter value) returns an empty result set
+  if first(DBInterface.execute(db, "PRAGMA integrity_check;")).integrity_check != "ok" ||
+     !(isempty(DBInterface.execute(db, "PRAGMA foreign_key_check;")))
+    throw(ErrorException("database integrity at risk or foreign key violation(s)"))
+  end
+
+  # According to https://www.geopackage.org/spec/#r10 and https://www.geopackage.org/spec/#r13
+  # A GeoPackage SHALL include a 'gpkg_spatial_ref_sys' table and a 'gpkg_contents table'
+  if first(DBInterface.execute(
+    db,
+    """
+  SELECT COUNT(*) AS n FROM sqlite_master WHERE 
+  name IN ('gpkg_spatial_ref_sys', 'gpkg_contents') AND 
+  type IN ('table', 'view');
+    """
+  )).n != 2
+    throw(ErrorException("missing required metadata tables in the GeoPackage SQL database"))
+  end
+
+  db
+end
+
+function iohandle(row, geomcolumn)
+  # get the column of SQL Geometry Binary specified by gpkg_geometry_columns table in column_name field
+  blob = getproperty(row, Symbol(geomcolumn))
+
+  # According to https://www.geopackage.org/spec/#r19
+  # A GeoPackage SHALL store feature table geometries in SQL BLOBs using the Standard GeoPackageBinary format
+  # check the GeoPackageBinaryHeader for the first byte[2] to be 'GP' in ASCII
+  if blob[1:2] != [0x47, 0x50]
+    @warn "Missing magic 'GP' string in GPkgBinaryGeometry"
+  end
+
+  # create in-memory I/O stream of GeoPackage SQL Geometry Binary Format
+  io = IOBuffer(blob)
+
+  # skip the magic string in header
+  seek(io, 3)
+
+  io
```

Comment:
I understand that you first perform a check with the raw blob of memory, create a IO handle, and then skip the blob related to the check. Could we rearrange the order of these operations to (1) create IO handle, (2) read from handle and perform check (that should already skip the memory chunk) and (3) return the IO handle?